File: MCC-01_algorithms-and-data-structures/algorithm-analysis/QuickSort/Python/indexQuickSort.py
import os
import time
import statistics
from quickSortPython import quickSort

print("*****************")
print("QUICK SORT PYTHON")
print("*****************")

def readArrayFromFile(filePath):
    try:
        # Este código utilizará la función open para abrir el archivo y luego leerá su contenido
        with open(filePath, 'r') as file:
            data = file.read()
            # Los elementos se separan por comas en el archivo de texto, por lo que se utilizan split(",") para separar los elementos y int(num.strip()) para convertir cada elemento en un número entero. 
            arr = [int(num.strip()) for num in data.split(",")]
            return arr
    except Exception as error:
        # Si ocurre algún error al leer el archivo, se mostrará un mensaje de error y se devolverá una lista vacía
        print("Error reading file:", error)
        return []

def processAndSortArray(filePath):
    unsortedArray = readArrayFromFile(filePath)

    numRepetitions = 10
    totalExecutionTime = 0
    sortedArray = []

    # Almacenar los tiempos individuales de ejecución en una lista en milisegundos
    executionTimes = []

    for i in range(1, numRepetitions + 1):
        # print("se repite", i, "veces")

        # usar time.perf_counter(), que devuelve en segundos, en lugar de time.time() para medir el tiempo transcurrido, proporciona una mayor precisión en la medición del tiempo y puede ser más adecuado para tareas de medición de rendimiento
        startTimeSort = time.perf_counter() * 1000 # Convertir a milisegundos
        sortedArray = quickSort(unsortedArray)
        endTimeSort = time.perf_counter() * 1000 # Convertir a milisegundos

        executionTime = (endTimeSort - startTimeSort) 
        executionTimes.append(executionTime)

        totalExecutionTime += (endTimeSort - startTimeSort)

    averageExecutionTime = totalExecutionTime / numRepetitions

    # Calcular la desviación estándar en milisegundos
    stdDeviation = statistics.stdev(executionTimes)

    print("--------------------")
    print(file_path)
    print("--------------------")

    print("Arreglo no ordenado:", unsortedArray)
    print("")
    print("Arreglo ordenado:", sortedArray)
    print("")
    print("Tiempo promedio de ejecución:", averageExecutionTime, "milisegundos")
    print("Desviación estándar:", stdDeviation, "milisegundos")
    print("********************************************************************************")

# Directorio donde se encuentra el script de Python
currentDirectory = os.path.dirname(os.path.abspath(__file__))

# Constante para hacer un formateo de strings o concatener de una forma especial
fileName = "../../Assets/unsortedArrays/";

# Ruta relativa al archivo "unsortedArray100.txt" (os.path.join construirá la ruta correcta basada en la ubicación actual del script de Python y la ruta proporcionada para el archivo)
unsorted_array_file_paths = [
  # os.path.join(currentDirectory, f"{fileName}unsortedArray100.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray1000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray2000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray3000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray4000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray5000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray6000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray7000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray8000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray9000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray10000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray20000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray30000.txt"),
  # os.path.join(currentDirectory, f"{fileName}unsortedArray40000.txt"),
  os.path.join(currentDirectory, f"{fileName}unsortedArray50000.txt"),
]

# Procesar y ordenar cada archivo
for file_path in unsorted_array_file_paths:
    processAndSortArray(file_path)

File: algorithms-and-data-structures/algorithm-analysis/QuickSort/Go/indexQuickSort.go
// tiene que colocar en la terminal: go run QuickSort/Go/indexQuickSort.go QuickSort/Go/quickSortGolang.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

func nameAlgorithmh() {
	fmt.Println("*************")
	fmt.Println("QUICK SORT GO")
	fmt.Println("*************")
}

// readArrayFromFile reads an integer array from the provided file path.
func readArrayFromFile(filePath string) ([]int, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	data := string(content)
	strNums := strings.Split(data, ",")
	var arr []int

	for _, strNum := range strNums {
		num, err := strconv.Atoi(strings.TrimSpace(strNum))
		if err != nil {
			return nil, err
		}
		arr = append(arr, num)
	}

	return arr, nil
}

// processAndSortArray lee y ordena el arreglo del archivo, y muestra los resultados.
func processAndSortArray(filePath string) {
	unsortedArray, err := readArrayFromFile(filePath)
	if err != nil {
			fmt.Println("Error leyendo el archivo:", err)
			return
	}

	// Número de repeticiones para hallar su promedio de tiempo de ejecución
	numRepetitions := 10
	var totalExecutionTime time.Duration
	var sortedArray []int

	// Guardar los tiempos individuales en un slice en milisegundos
	var executionTimes []float64

	for i := 1; i <= numRepetitions; i++ {
			// fmt.Println("se repite", i, "veces")
			startTimeSort := time.Now()
			sortedArray = quickSort(unsortedArray)
			endTimeSort := time.Now()

			executionTime := endTimeSort.Sub(startTimeSort).Seconds() * 1000 // Convertir a milisegundos
			executionTimes = append(executionTimes, executionTime)

			totalExecutionTime += endTimeSort.Sub(startTimeSort)
	}

	averageExecutionTime := totalExecutionTime / time.Duration(numRepetitions)

	// Calcular la desviación estándar en milisegundos
	var sum float64
	for _, time := range executionTimes {
			sum += time
	}
	mean := sum / float64(len(executionTimes))

	var variance float64
	for _, time := range executionTimes {
			variance += (time - mean) * (time - mean)
	}
	variance = variance / float64(len(executionTimes))

	stdDeviation := math.Sqrt(variance)

	fmt.Println("--------------------")
	fmt.Println(filePath)
	fmt.Println("--------------------")

	fmt.Println("Arreglo no ordenado:", unsortedArray)
	fmt.Println("")
	fmt.Println("Arreglo ordenado:", sortedArray)
	fmt.Println("")
	fmt.Println("Tiempo promedio de ejecución:", averageExecutionTime.Milliseconds(), "milisegundos")
	fmt.Println("Desviación estándar:", stdDeviation, "milisegundos")
	fmt.Println("********************************************************************************")
}

func main() {
	nameAlgorithmh()
	
	// Obtener el directorio actual.
	currentDirectory, err := os.Getwd()
	if err != nil {
			fmt.Println("Error obteniendo el directorio actual:", err)
			return
	}

  // Obtener la ruta absoluta de la carpeta ASSETS
  assetsPath := filepath.Join(currentDirectory, "ASSETS")

	// Definir las rutas de los archivos.
	filePaths := []string{
		filepath.Join(assetsPath, "unsortedArrays", "unsortedArray100.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray1000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray2000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray3000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray4000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray5000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray6000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray7000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray8000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray9000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray10000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray20000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray30000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray40000.txt"),
		// filepath.Join(assetsPath, "unsortedArrays", "unsortedArray50000.txt"),
	}

	// Procesar y ordenar cada archivo
	for _, file := range filePaths {
		processAndSortArray(file)
}
}